Rubberduck.VBEEditor/ComManagement/TypeLibs.cs

```diff
@@ -0,0 +1,707 @@
+
+        private void CacheCommonProperties()
+        {
+            IntPtr typeAttrPtr = IntPtr.Zero;
+            GetTypeAttr(out typeAttrPtr);
+            _cachedAttributes = StructHelper.ReadStructure<ComTypes.TYPEATTR>(typeAttrPtr);
+            ReleaseTypeAttr(typeAttrPtr);      // don't need to keep a hold of it, as _cachedAttributes is a copy
+
+            // cache the container type library if it is available
+            try
+            {
+                // We have to wrap the ITypeLib returned by GetContainingTypeLib
+                // so we cast to our ITypeInfo_Ptrs interface in order to work with the raw IntPtrs
+                IntPtr typeLibPtr = IntPtr.Zero;
+                _ITypeInfoAlt.GetContainingTypeLib(out typeLibPtr, out _containerTypeLibIndex);
+                _containerTypeLib = new TypeLibWrapper(typeLibPtr);  // takes ownership of the COM reference
+            }
+            catch (Exception)
+            {
+                // it is acceptable for a type to not have a container, as types can be runtime generated.
+            }
+
+            if (Name == null) _wrappedObject.GetDocumentation((int)TypeLibConsts.MEMBERID_NIL, out _name, out _docString, out _helpContext, out _helpFile);
+        }
+
+        public TypeInfoWrapper(ComTypes.ITypeInfo rawTypeInfo)
+        {
+            _wrappedObject = rawTypeInfo;
+            CacheCommonProperties();
+        }
+
+        public TypeInfoWrapper(IntPtr rawObjectPtr, int? parentUserFormUniqueId = null)
+        {
+            _typeInfoAggregatorObj = new AggregateInterfacesWrapper(rawObjectPtr, new Type[] { typeof(ComTypes.ITypeInfo), typeof(ITypeInfo_VBE) });
+            _wrappedObject = (ComTypes.ITypeInfo)_typeInfoAggregatorObj.WrappedObject;
+            
+            // base classes of VBE UserForms cause an access violation on calling GetDocumentation(MEMBERID_NIL)
+            // so we have to detect UserForm parents, and ensure GetDocumentation(MEMBERID_NIL) never gets through
+            if (parentUserFormUniqueId.HasValue)
+            {
+                _name = "_UserFormBase{unnamed}#" + parentUserFormUniqueId;
+            }
+
+            CacheCommonProperties();
+            DetectUserFormClass();
+        }
+
+        public bool HasPredeclaredId { get => _cachedAttributes.wTypeFlags.HasFlag(ComTypes.TYPEFLAGS.TYPEFLAG_FPREDECLID); }
+
+        private bool IsRuntimeGenerated() => _containerTypeLib == null;
+
+        private void DetectUserFormClass()
+        {
+            // Determine if this is a UserForm base class, that requires special handling to workaround a VBE bug in its implemented classes
+            // the guids are dynamic, so we can't use them for detection.
+            if ((_cachedAttributes.typekind == ComTypes.TYPEKIND.TKIND_COCLASS) &&
+                    IsVBEHosted() &&
+                    IsRuntimeGenerated() &&
+                    (_cachedAttributes.cImplTypes == 2) && (Name == "Form"))
```

Comment:
Also, in the method `GetUserFormControlType `, you check if it has `FormItf`. Can we check that here, too?

Reply:
The 2 check should be fine.  This is one of the base generated classes, whereas an Implements should occur higher up the food chain.  I'll double check that though.

Reply:
IIRC, the FormItf interface is off a different branch of the UserForm implementations, so can't be verified without somewhat more effort.

Reply:
I've checked this, and it is fine.   Actually, any 'Implements' statements do not reach the type library at all, so that's another quirk to be aware of.  As with enums/UDTs, it's probably fixable if the need arises.

Reply:
WRT FormItf:  each UserForm implements two interfaces, "FormItf" and "Form".  It is the "Form" one that further implements the two rogue interfaces.  Once we've got to the "Form" typeinfo we no longer have information on the superclass UserForm to check for FormItf.  We could possibly move the detection of the UserForm closer to the superclass, but I suspect that would actually be a lot less efficient, and we wouldn't gain much benefit at all.  The IsRuntimeGenerated call in the detection is actually pretty damn reliable here, and has been renamed more appropriately to HasNoContainer in my local branch.

Reply:
Hmm. That uncovered a mistaken assumption on my part - I expected that when testing for interfaces, I would get all interfaces returned including the base class' interfaces, but that's not the case here given a `ITypeInfo` interface -- IOW, we can't cast a `ITypeInfo` interface to its base class `MSForms.Form` or its derivatives, correct? I'm mainly noting that because it's too easy to forget that in COM world, not all hierarchies of implemented interfaces are exposed compared to .NET world.

Reply:
Yeah, an ITypeInfo only exposes directly implemented interfaces of that exact class/interface.  You have to recursively go through all base classes to get all interfaces.   My DoesImplement() implementation does search all base interfaces recursively.